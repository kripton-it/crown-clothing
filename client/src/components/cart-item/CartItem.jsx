import React from "react";

import {
  CartItemContainer,
  ItemsDetails,
  ItemName,
  ItemPrice
} from "./CartItemStyles";

const CartItem = ({ item }) => {
  const { imageUrl, price, name, quantity } = item;
  return (
    <CartItemContainer>
      <img src={imageUrl} alt={name} />
      <ItemsDetails>
        <ItemName>{name}</ItemName>
        <ItemPrice>{`${quantity} x $${price}`}</ItemPrice>
      </ItemsDetails>
    </CartItemContainer>
  );
};

export default CartItem;
