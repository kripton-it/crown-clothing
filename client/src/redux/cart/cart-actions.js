import CartActionTypes from "./cart-action-types";

const {
  TOGGLE_CART_HIDDEN,
  ADD_ITEM,
  REMOVE_ITEM,
  DECREASE_AMOUNT,
  CLEAR_CART
} = CartActionTypes;

// переключение корзины (открыта/закрыта)
export const toggleCartHiddenAction = () => ({
  type: TOGGLE_CART_HIDDEN
});

// добавление товара в корзину
export const addItemToCartAction = item => ({
  type: ADD_ITEM,
  payload: item
});

// удаление товара из корзины
export const removeItemFromCartAction = item => ({
  type: REMOVE_ITEM,
  payload: item
});

// уменьшение количества товара на 1
export const decreaseAmountAction = item => ({
  type: DECREASE_AMOUNT,
  payload: item
});

// очистка корзины
export const clearCart = () => ({
  type: CLEAR_CART
});