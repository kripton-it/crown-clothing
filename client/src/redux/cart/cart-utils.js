// добавление нового товара в список товаров с группировкой одинаковых товаров
export const addItemToCart = (items, newItem) => {
  const existingItem = items.find(item => item.id === newItem.id);
  return existingItem
    ? items.map(item =>
        item.id === newItem.id ? { ...item, quantity: item.quantity + 1 } : item
      )
    : [...items, { ...newItem, quantity: 1 }];
};

// уменьшение количества товара на 1
export const decreaseItemAmount = (items, itemToDecreaseAmount) =>
  itemToDecreaseAmount.quantity === 1
    ? items.filter(item => item.id !== itemToDecreaseAmount.id)
    : items.map(item =>
        item.id === itemToDecreaseAmount.id
          ? { ...item, quantity: item.quantity - 1 }
          : item
      );
