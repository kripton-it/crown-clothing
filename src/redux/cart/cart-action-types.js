const CartActionTypes = {
  TOGGLE_CART_HIDDEN: "TOGGLE_CART_HIDDEN",
  ADD_ITEM: "ADD_ITEM",
  REMOVE_ITEM: "REMOVE_ITEM",
  DECREASE_AMOUNT: "DECREASE_AMOUNT"
}

export default CartActionTypes;