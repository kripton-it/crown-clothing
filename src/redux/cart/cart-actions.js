import CartActionTypes from "./cart-action-types";

const { TOGGLE_CART_HIDDEN, ADD_ITEM, REMOVE_ITEM } = CartActionTypes;

export const toggleCartHiddenAction = () => ({
  type: TOGGLE_CART_HIDDEN
});

export const addItemToCartAction = item => ({
  type: ADD_ITEM,
  payload: item
});

export const removeItemFromCartAction = id => ({
  type: REMOVE_ITEM,
  payload: id
});
