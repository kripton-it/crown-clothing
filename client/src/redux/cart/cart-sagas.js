import { takeLatest, put, all, call, select } from "redux-saga/effects";
import UserActiontypes from "../user/user-action-types";
import CartActiontypes from "./cart-action-types";

import { clearCart, setCartFromFirebase } from "./cart-actions";

import { selectCurrentUser } from "../user/user-selectors";
import { selectCartItems } from "./cart-selectors";

import { getUserCartRef } from "../../firebase/firebase.utils";

const { SIGN_OUT_SUCCESS, SIGN_IN_SUCCESS } = UserActiontypes;
const { ADD_ITEM, REMOVE_ITEM, DECREASE_AMOUNT, CLEAR_CART } = CartActiontypes;

export function* clearCartAfterSignOutSuccess() {
  yield put(clearCart());
}

export function* updateCartInFirebase() {
  const currentUser = yield select(selectCurrentUser);
  if (currentUser) {
    try {
      const cartRef = yield getUserCartRef(currentUser.id);
      const cartItems = yield select(selectCartItems);
      yield cartRef.update({ cartItems });
    } catch (error) {
      console.log(error);
    }
  }
}

export function* checkCartFromFirebase({ payload: user }) {
  const cartRef = yield getUserCartRef(user.id);
  const cartSnapshot = yield cartRef.get();
  yield put(setCartFromFirebase(cartSnapshot.data().cartItems));
}

export function* signOutSuccess() {
  yield takeLatest(SIGN_OUT_SUCCESS, clearCartAfterSignOutSuccess);
}

export function* onUserSignIn() {
  yield takeLatest(SIGN_IN_SUCCESS, checkCartFromFirebase);
}

export function* onCartChange() {
  yield takeLatest(
    [ADD_ITEM, REMOVE_ITEM, DECREASE_AMOUNT, CLEAR_CART],
    updateCartInFirebase
  );
}

export default function* cartSagas() {
  yield all([call(signOutSuccess), call(onCartChange), call(onUserSignIn)]);
}
