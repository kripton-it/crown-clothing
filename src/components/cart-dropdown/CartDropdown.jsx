import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import { selectCartItems } from "../../redux/cart/cart-selectors";

import "./CartDropdown.styles.scss";

import CartItem from "../cart-item/CartItem";

import CustomButton from "../custom-button/CustomButton";

const CartDropdown = ({ cartItems }) => {
  return (
    <div className="cart-dropdown">
      <div className="cart-items">
        {cartItems.map(cartItem => (
          <CartItem key={cartItem.id} item={cartItem} />
        ))}
        {/* {cartItems.length ? cartItems.map(cartItem => (
          <CartItem key={cartItem.id} item={cartItem} />
        )) : <span classname="empty-message">Your cart is empty</span>} */}
      </div>
      <CustomButton>Go to checkout</CustomButton>
      {/*{cartItems.length ? <CustomButton>Go to checkout</CustomButton> : null} */}
    </div>
  );
};

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems
});

export default connect(mapStateToProps)(CartDropdown);
