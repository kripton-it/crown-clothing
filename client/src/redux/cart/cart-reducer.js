import CartActionTypes from "./cart-action-types";

import { addItemToCart, decreaseItemAmount } from "./cart-utils";

const INITIAL_STATE = {
  hidden: true,
  cartItems: []
};

const {
  TOGGLE_CART_HIDDEN,
  ADD_ITEM,
  REMOVE_ITEM,
  DECREASE_AMOUNT,
  CLEAR_CART
} = CartActionTypes;

const cartReducer = (state = INITIAL_STATE, action) => {
  const { hidden, cartItems } = state;
  switch (action.type) {
    case TOGGLE_CART_HIDDEN:
      return {
        ...state,
        hidden: !hidden
      };
    case ADD_ITEM:
      return {
        ...state,
        cartItems: addItemToCart(cartItems, action.payload)
      };
    case REMOVE_ITEM:
      return {
        ...state,
        cartItems: cartItems.filter(({ id }) => id !== action.payload.id)
      };
    case DECREASE_AMOUNT:
      return {
        ...state,
        cartItems: decreaseItemAmount(cartItems, action.payload)
      };
    case CLEAR_CART:
      return {
        ...state,
        cartItems: []
      };
    default:
      return state;
  }
};

export default cartReducer;
