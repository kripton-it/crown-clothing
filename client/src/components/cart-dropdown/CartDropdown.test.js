import React from "react";
import { shallow } from "enzyme";

import CartDropdown from "./CartDropdown";
import CartItem from '../cart-item/CartItem';

import { toggleCartHiddenAction } from '../../redux/cart/cart-actions';

describe("CartDropdown component", () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }];

  beforeEach(() => {
    mockHistory = {
      push: jest.fn()
    }

    mockDispatch = jest.fn()

    const mockProps = {
      history: mockHistory,
      dispatch: mockDispatch,
      cartItems: mockCartItems
    };

    wrapper = shallow(<CartDropdown {...mockProps} />)
  })

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call history.push when button is clicked', () => {
    wrapper.find('CustomButton').simulate('click');
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHiddenAction());
  });

  it('should render an equal number of CartItem components as the cartItems prop', () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });
})

