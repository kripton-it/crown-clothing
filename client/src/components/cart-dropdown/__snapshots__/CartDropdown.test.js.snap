// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component should render 1`] = `ShallowWrapper {}`;
